designs/spot-consolidation.md

```diff
@@ -0,0 +1,105 @@
+
+Karpenter’s consolidation algorithm minimizes the price of EC2 instances given pod scheduling constraints. However, price minimization presents challenges for EC2 Spot, which offers a tradeoff between price and interruption rate. Karpenter uses EC2 Spot’s PriceCapacityOptimized (PCO) allocation strategy to provide a balanced approach to this tradeoff, and doesn't always select the cheapest instance type, since it may result in immediate interruption. If Karpenter were to consolidate spot capacity, it would repeatedly interrupt the same instance, walking down the PCO decision ladder, until only the lowest price remained. Because of these complications, Karpenter does not currently support "spot replacement" consolidation.
+
+All consolidation actions contain an inherent tradeoff between price and availability. Consolidation actions are equivalent to a spot interruption, but without the deadline. This tradeoff applies to on-demand capacity, as well. This document explores approaches to enable consolidation for spot instances without regressing to lowest price. 
+
+## Design Options
+
+### 1: Minimum Flexibility [Recommended] 
+
+Karpenter could skip consolidation if the number of potential replacement instance types were below a threshold (e.g. 20). To avoid repeat interruptions, e.g. "walking down the ladder", Karpenter would have to limit its flexibility to this threshold. Further, Karpenter would have to limit its flexibility during the launch scenario as well, else face an immediate consolidation if the launched instance type’s price were more than the price of any instance type under the flexibility threshold. Since on-demand capacity uses the lowest price strategy, its behavior is unaffected if these rules are applied equally to both spot and on-demand launches.
+
+Conceptually, this approach is straightforward to explain to customers and directly aligned with Spot best practices. Some customers may be disappointed by the flexibility requirement, but given the current lack of support for spot consolidation, it’s net positive from the status quo. While there may be valid use cases for customers to trade interruption rate for lower prices, we should delay enabling customers to configure this value until the use cases are better understood. 
+
+* 👍👍👍 Prevents consolidation from regressing to a lowest price allocation strategy
+* 👍 Easy to explain and implements spot best practices
+* 👍 Flexibility threshold is tunable, as best practices evolve. Cloud providers could choose to set it to 1 to disable this behavior.
+* 👎 Limits potential flexibility from status quo for both instance launch and replacement. Today, Karpenter can send unbounded flexibility to PCO (e.g. 100s of instance types). 
+* 👎 Fails to take advantage of opportunities to migrate to healthier pools due to limited instance type flexibility.
```

Comment:
This may be captured within this item, but it also doesn't allow us to consolidate to the "best" overall pool in all cases. This effectively causes us to get to a state of good enough that accounts for us getting probably 90% of the way there in cost reduction

Reply:
That's exactly the intent of this con. Is there something about the language I can make more clear?

Reply:
I didn't see cost mentioned here. From "healthy" I was understanding availability, but I think cost improvements are also constrained here too

Reply:
@nikmohan123 Can you expound on this item here?

Reply:
Addressed.